[DONOTMERGE] Switch llvm source build to v20

Zentrik opened the pull request:
This is for testing with pkgeval before we have built on yggdrasil

@nanosoldier `runtests(configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1", "USE_BINARYBUILDER_LLVM=0"],), vs_configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"],))`

Zentrik committed:
[DONOTMERGE] Switch llvm source build to v20

Zentrik committed:
Bump to rebased llvm 20

Zentrik committed:
Cherry-pick "Fix late gc lowering pass for vector intrinsics #55864"

maleadt commented:
PkgEval got restarted:

@nanosoldier `runtests(configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1", "USE_BINARYBUILDER_LLVM=0"],), vs_configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"],))`

maleadt commented:
PkgEval got restarted:

@nanosoldier `runtests(configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1", "USE_BINARYBUILDER_LLVM=0"],), vs_configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"],))`

Only one job ahead of this, so it should report within 2 days.

maleadt commented:
The build failed:

```
CMake 3.20.0 or higher is required.  You are running version 3.18.4
```

PkgEval is currently using the latest rootfs, https://github.com/JuliaCI/rootfs-images/releases/tag/v7.10. I guess we'll need to upgrade from bullseye to bookworm, https://github.com/JuliaCI/rootfs-images/blob/e7be8086ff1285b0f333ca72884977d01aed0e1f/src/build_img/debian.jl#L17.

giordano commented:
You can download cmake with [`contrib/download_cmake.sh `](https://github.com/JuliaLang/julia/blob/7d84c5369b869eec22a9478e0ea492488e880e6d/contrib/download_cmake.sh), but no idea how do that for pkgeval invocations (_if_ doable at all)

nanosoldier commented:
[Your job](https://github.com/JuliaLang/julia/pull/57658#issuecomment-2708698367) failed. Consult the server logs for more details (cc @maleadt).

Zentrik commented:
I forgot we added an invocation to the download cmake script for the source build, https://github.com/JuliaCI/julia-buildkite/commit/552037be10cf2e4dea4b61eaacc790b5bc176f6b.

Zentrik commented:
@maleadt would it be easy to get this fixed otherwise I'll start the llvm yggy build?

giordano commented:
Can't you run the download cmake script as part of the build script in this PR?  This is just for testing anyway.

Zentrik commented:
> Can't you run the download cmake script as part of the build script in this PR? This is just for testing anyway.

Yep, ignore me then tim, I'll do that.

Zentrik committed:
Try bumping cmake before building llvm

Zentrik commented:
Let's see if this works

@nanosoldier `runtests(configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1", "USE_BINARYBUILDER_LLVM=0"],), vs_configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"],))`

Zentrik committed:
Try building from source on ci

Zentrik committed:
try again

Zentrik committed:
try try again

Zentrik committed:
try try try again

Zentrik committed:
Update Makefile

Zentrik committed:
Adapt to "[ORC] Move EHFrameRegistrationPlugin into its own header + source file. NFC.
"

https://github.com/llvm/llvm-project/commit/6d12b954a7df11e32acf110950d88bac282079b8

Zentrik committed:
Fix deprecation

Zentrik committed:
Add another override

Zentrik committed:
Fix if on unsupported platform

Zentrik committed:
Hack

Zentrik committed:
Fix bug

Zentrik committed:
Try Fix bug

Zentrik committed:
Try again

Zentrik committed:
Try again

Zentrik committed:
Fix typo

Zentrik committed:
Fix another deprecation

Zentrik committed:
Fix more deprecations

Zentrik committed:
Fix another deprecation

Zentrik committed:
hopefully last one

Zentrik committed:
Remove release banner override

Zentrik committed:
Fix more deprecations

Zentrik commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -2544,7 +2579,7 @@ void LateLowerGCFrame::PlaceRootsAndUpdateCalls(ArrayRef<int> Colors, int PreAss
                     getOrDeclare(jl_intrinsics::popGCFrame),
                     {gcframe});
-                popGcframe->insertBefore(BB.getTerminator());
+                popGcframe->insertBefore(BB.getTerminator().getIterator());
```
```suggestion
                popGcframe->insertBefore(BB.getTerminator()->getIterator());
```

Zentrik committed:
Update src/llvm-late-gc-lowering.cpp

nanosoldier commented:
[Your job](https://github.com/JuliaLang/julia/pull/57658#issuecomment-2710647868) failed. Consult the server logs for more details (cc @maleadt).

Zentrik committed:
Include checksum and remove hack

Zentrik commented:
Hmm, this just seems like a llvm bug where it's not specifying the path to zlib.

Zentrik committed:
Hack to get llvm to link against correct zlib

Filed https://github.com/llvm/llvm-project/issues/131119

Zentrik commented on deps/llvm.mk:
```diff
@@ -69,7 +73,7 @@ LLVM_EXPERIMENTAL_TARGETS :=
 LLVM_CFLAGS :=
 LLVM_CXXFLAGS :=
-LLVM_CPPFLAGS :=
+LLVM_CPPFLAGS := -L"$(build_prefix)/lib/libz.so"
```
```suggestion
LLVM_CPPFLAGS := -L"$(build_prefix)/lib/libz.so.1"
```

Zentrik committed:
Update deps/llvm.mk

Zentrik commented on deps/llvm.mk:
```diff
@@ -69,7 +73,7 @@ LLVM_EXPERIMENTAL_TARGETS :=
 LLVM_CFLAGS :=
 LLVM_CXXFLAGS :=
-LLVM_CPPFLAGS :=
+LLVM_CPPFLAGS := -L"$(build_prefix)/lib/libz.so.1"
```
```suggestion
LLVM_CPPFLAGS := "$(build_prefix)/lib/libz.so.1"
```

Zentrik committed:
Update deps/llvm.mk

Zentrik commented:
Let's try this one more time

@nanosoldier `runtests(configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1", "USE_BINARYBUILDER_LLVM=0"],), vs_configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"],))`

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/57658#issuecomment-2721560696) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/3d11791_vs_a001847/report.html) is available.

<details><summary>Report summary</summary>

## ❗ Packages that crashed

**2 packages crashed only on the current version.**

 - A segmentation fault happened: 2 packages


## ✖ Packages that failed

**7 packages failed only on the current version.**

 - Package fails to precompile: 1 packages
 - Package has test failures: 3 packages
 - Test duration exceeded the time limit: 2 packages
 - Test log exceeded the size limit: 1 packages

673 packages failed on the previous version too.

## ✔ Packages that passed tests

**10 packages passed tests only on the current version.**

 - Other: 10 packages

3665 packages passed tests on the previous version too.

## ~ Packages that at least loaded

**1 packages successfully loaded only on the current version.**

 - Other: 1 packages

1987 packages successfully loaded on the previous version too.

## ➖ Packages that were skipped altogether

4034 packages were skipped on the previous version too.



</details>

giordano commented:
That's remarkably few new failures.

Zentrik committed:
Bump to latest llvm

Zentrik committed:
Add checksums

Zentrik commented:
All looked good, but let's confirm on the newer commit from our llvm fork. 

@nanosoldier `runtests(configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1", "USE_BINARYBUILDER_LLVM=0"],), vs_configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"],))`

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/57658#issuecomment-2725123657) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1b962d6_vs_b8dc57e/report.html) is available.

<details><summary>Report summary</summary>

## ❗ Packages that crashed

**2 packages crashed only on the current version.**

 - A segmentation fault happened: 2 packages


## ✖ Packages that failed

**7 packages failed only on the current version.**

 - Package has test failures: 4 packages
 - Networking-related issues were detected: 1 packages
 - Test duration exceeded the time limit: 2 packages

606 packages failed on the previous version too.

## ✔ Packages that passed tests

**4 packages passed tests only on the current version.**

 - Other: 4 packages

3334 packages passed tests on the previous version too.

## ~ Packages that at least loaded

1737 packages successfully loaded on the previous version too.

## ➖ Packages that were skipped altogether

4691 packages were skipped on the previous version too.



</details>

Zentrik committed:
Update test

IanButterworth commented:
Good to rerun now

Zentrik committed:
minor fixes

Zentrik commented:
@nanosoldier `runtests(configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1", "USE_BINARYBUILDER_LLVM=0"],), vs_configuration = (buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"],))`

Zentrik commented:
Hmm, this x86 darwin failure looks real given it doesn't occur on https://github.com/JuliaLang/julia/pull/57792.
```
 clang++ -mmacosx-version-min=10.14 -m64  -shared -fPIC -L[buildroot]/usr/lib/julia -L[buildroot]/usr/lib -L[buildroot]/usr/lib -o [buildroot]/usr/lib/julia/sysbase.dylib -Xlinker -all_load [buildroot]/usr/lib/julia/sysbase-o.a  -ljulia-internal -ljulia $([ Darwin = WINNT ] && echo '' -lopenlibm -lssp --disable-auto-import --disable-runtime-pseudo-reloc)
duplicate symbol '___extendhfsf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#1.o)
duplicate symbol '___truncsfhf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#1.o)
duplicate symbol '___truncdfhf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#1.o)
duplicate symbol '___extendhfsf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#2.o)
duplicate symbol '___truncsfhf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#2.o)
duplicate symbol '___extendhfsf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#3.o)
duplicate symbol '___truncsfhf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#3.o)
duplicate symbol '___truncdfhf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#3.o)
duplicate symbol '___extendhfsf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#4.o)
duplicate symbol '___truncsfhf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#4.o)
duplicate symbol '___extendhfsf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#5.o)
duplicate symbol '___truncsfhf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#5.o)
duplicate symbol '___extendhfsf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#6.o)
duplicate symbol '___truncsfhf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#6.o)
duplicate symbol '___truncdfhf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#6.o)
duplicate symbol '___extendhfsf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#7.o)
duplicate symbol '___truncsfhf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#7.o)
duplicate symbol '___truncdfhf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#7.o)
duplicate symbol '___extendhfsf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#8.o)
duplicate symbol '___truncsfhf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#8.o)
duplicate symbol '___extendhfsf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#9.o)
duplicate symbol '___truncsfhf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#9.o)
duplicate symbol '___extendhfsf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#10.o)
duplicate symbol '___truncsfhf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#10.o)
duplicate symbol '___extendhfsf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#11.o)
duplicate symbol '___truncsfhf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#11.o)
duplicate symbol '___truncdfhf2' in:
    [buildroot]/usr/lib/julia/sysbase-o.a(text#0.o)
    [buildroot]/usr/lib/julia/sysbase-o.a(text#11.o)
ld: 27 duplicate symbols for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [[buildroot]/usr/lib/julia/sysbase.dylib] Error 1
```

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/57658#issuecomment-2727095582) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/444b00a_vs_04922e7/report.html) is available.

<details><summary>Report summary</summary>

## ❗ Packages that crashed

**2 packages crashed only on the current version.**

 - An internal error was encountered: 1 packages
 - A segmentation fault happened: 1 packages

1 packages crashed on the previous version too.

## ✖ Packages that failed

**19 packages failed only on the current version.**

 - Package has test failures: 6 packages
 - Package tests unexpectedly errored: 1 packages
 - Test duration exceeded the time limit: 11 packages
 - Test log exceeded the size limit: 1 packages

1449 packages failed on the previous version too.

## ✔ Packages that passed tests

**20 packages passed tests only on the current version.**

 - Other: 20 packages

5049 packages passed tests on the previous version too.

## ~ Packages that at least loaded

**16 packages successfully loaded only on the current version.**

 - Other: 16 packages

2923 packages successfully loaded on the previous version too.

## ➖ Packages that were skipped altogether

**1 packages were skipped only on the current version.**

 - Package could not be installed: 1 packages

905 packages were skipped on the previous version too.



</details>
